corehq/apps/api/resources/v0_5.py

```diff
@@ -382,24 +371,29 @@ def obj_create(self, bundle, request=None, **kwargs):
                 _set_role_for_bundle(kwargs, bundle)
             bundle.obj.save()
         except Exception:
-            bundle.obj.delete()
+            if bundle.obj._id:
+                bundle.obj.delete(deleted_by=bundle.request.user, deleted_via=USER_CHANGE_VIA_API)
+            else:
+                try:
+                    django_user = bundle.obj.get_django_user()
+                except User.DoesNotExist:
+                    pass
+                else:
+                    django_user.delete()
+                    log_model_change(bundle.request.user, django_user, message=f"deleted_via: {USER_CHANGE_VIA_API}",
+                                     action=ModelAction.DELETE)
+            raise
```

Comment:
This is mostly copied from the CommCareUserResource. Perhaps it could be extracted to a utility